### ISSUE TITLE
Fillet can not round a face

### ISSUE THREAD

**Issue by farahats9** [Open-Cascade-SAS/OCCT#172]
### Category

Modeling

### Severity

Major

### Description

I know this bug is in the original tracker [25478](https://tracker.dev.opencascade.org/view.php?id=25478)
but it has been 10 years since it was reported and there is no sign of any attention or progress done on this. Please someone needs to look into it since it is feels like the most important problem now in FreeCAD

### Expected Behavior

if the fillet radius is big enough that there are no linear faces remaining, it should round the face.

### Actual Behavior

The fillet will work as long as the radius is small enough to keep some face remaining. if it is big it will not work.

### Sample Code or DRAW Tcl Script

see original issue for more description and examples. [25478](https://tracker.dev.opencascade.org/view.php?id=25478)

### Operating System

Linux

### Compiler

GCC

### Bitness

64-bit

### OCCT Version

latest

**Comment by GeneGH** [comment]
What do you mean by "round the face"? 

* Use the requested radius and leave sharp corners?

* Change the radius to fit?

* Just create some artibrary shape that looks "round"?

**Comment by farahats9** [comment]
It may be easier to explain visually:

I have the following simple shape with height 10 mm
![image](https://github.com/user-attachments/assets/44a17489-8804-49bf-8dea-5d447a548403)

now I want to apply a fillet on all sides that is 4 mm, will work without any problem
![image](https://github.com/user-attachments/assets/50e5c92a-cbe1-48d3-99ee-8751f3a30dec)


because there is still a face connecting the 2 red fillets there is no problem so far.

now if I want to make the fillet 5mm that means they will touch and the grey face between them will have 0 mm length and will refuse to do the fillet with the error:
```BRep_API: command not done```

I hope this makes the issue a bit clearer, this is also in **chamfer** not just a fillet problem, let me know if you need any information or if I can be of any help.

**Comment by GeneGH** [comment]
I understand the issue. I have seen it many times.

The question is what should happen when the fillet cannot be completed as specified.

**Comment by farahats9** [comment]
I don't know if it will be easy but the logic of this should be: If the fillet causes a face dimension to be 0 then delete that face.

What should happen in my example is the gray face that was 2 mm and became 0 mm due to fillets should be deleted and **the upper and bottom fillets should connect as tangents**. Does that make sense? I don't know if there is an easier way to make this work but that's how it is in other commercial CAD software.

**Comment by GeneGH** [comment]
OK, now they are tangent at the former face along the edge. What happens at the other ends? Sharp break in the curve? Changed radius of the fillet so that the other ends can be tangent to those faces?

**Comment by farahats9** [comment]
In my example both the top and bottom fillets have the same radius so both of them would be tangent without changing any radius.

But if I understand you correctly you are asking what if they cannot be tangent due to radius difference or other constraints, from my experience in other CAD software, you should never change the other fillet's radius, keep the sharp corners and let the dimension changes in the user's control.

**Comment by dragazo** [comment]
I don't know if this would help in the implementation, but I believe the behavior used by commercial CAD software is equivalent to performing the fillet on each edge loop separately and then taking the intersection. So in this example we would duplicate the part, perform the top fillet on one copy, perform the bottom fillet on the other copy, and then do a boolean intersection to get the final result. As long as boolean intersections don't have some other issue, this should work for both over-fillet and critical-fillet (as is used in this issue), so two birds with one stone.

**Comment by drwho495** [comment]
any updates on this?

**Comment by MisterMakerNL** [comment]
This one bug is pretty bad, with the Naming issue the fillet can cause this one is probably one of the biggest issues there are currenty for using OCCT for a CAD program. Are there any plans on working on this?

**Comment by angelobartolome** [comment]
I've manually tested @dragazo's suggestion on a simpler version, filleting two edges until they overlap, seems quite promising

<img width="516" height="482" alt="Image" src="https://github.com/user-attachments/assets/ab2b1d87-124d-4433-bd40-553282606eed" />

Interestingly, the concave example provided doesn't quite work in some other commercial CAD applications, such as OnShape.

<img width="931" height="586" alt="Image" src="https://github.com/user-attachments/assets/a095586c-4e4f-4c7c-862d-b69e2c9ab50e" />

However it does currently work in OCCT if you do the outer edges first then the middle one, so I think this should not be the focus for the issue for now.

And here is a "working" example for the "fillet" face reported in https://github.com/Open-Cascade-SAS/OCCT/issues/172#issuecomment-2508194883.

<img width="761" height="539" alt="Image" src="https://github.com/user-attachments/assets/da871b90-aad2-4a65-8065-870db00cdbdb" />

I've built a test naive-API that will do one fillet at a time then run the intersection, it works for these use cases but it is way slower than the previous approach (all-in-one). 

Maybe we can surface the `OneCorner : fillets have too big radiuses` exception that is thrown on an higher level in the API and then execute this naive approach.

**Comment by angelobartolome** [comment]
If anyone wants to test in FreeCAD, I'm sharing my test API here, I'll not open an PR for this because it is very simple and probably not adequate solution, since I'm pretty much duplicating the FilletAPI adding that feature. I think the best course would be to discuss how to surface this API, if we would extend the original `BRepFilletAPI_MakeFillet`, or something else.

https://github.com/angelobartolome/OCCT/commit/1b81bbbcc420ded437e347578e570b543853afb2

**Comment by farahats9** [comment]
I understand why you don't want to add the solution in a PR in its current form but if we leave it for discussion it will be years before it is implemented, I recommend adding the PR since it is already better than what we have.

Also since Chamfers have the same exact problem do you think you can apply a similar fix to it?

**Comment by drwho495** [comment]
does this have the Modified and Generated methods? freecad has to have those for its toponaming algorithm

**Comment by MisterMakerNL** [comment]
Really nice work also the solution sounds very simple! 
Thanks for looking into this!

**Comment by pierreporte** [comment]
In any case, this concave situation wouldn’t be fixed, right? (1) is the geometry before applying the filet, (2) is when a fillet with radius higher than the step is applied.

<img width="830" height="611" alt="Image" src="https://github.com/user-attachments/assets/850bfa4d-8cdc-4289-a0bc-c9a88b724760" />

**Comment by angelobartolome** [comment]
> does this have the Modified and Generated methods? freecad has to have those for its toponaming algorithm
I think should yes 🤔, haven't tried yet since I'm doing my tests using the Draw Test Harness (Tcl scripts)


> In any case, this concave situation wouldn’t be fixed, right? (1) is the geometry before applying the filet, (2) is when a fillet with radius higher than the step is applied.
> 
> <img alt="Image" width="830" height="611" src="https://private-user-images.githubusercontent.com/2029853/482741941-850bfa4d-8cdc-4289-a0bc-c9a88b724760.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYzMTE1NDgsIm5iZiI6MTc1NjMxMTI0OCwicGF0aCI6Ii8yMDI5ODUzLzQ4Mjc0MTk0MS04NTBiZmE0ZC04Y2RjLTQyODktYTBiYy1jOWE4OGI3MjQ3NjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MjdUMTYxNDA4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjcwNjUyMDY2NTkyOTNmYWVhMmNiMjUyNzRlMzIzNWM2N2I2Y2ZjZThmNDc3ZjExMDZlYWZhNTVjY2MzNTU4YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.3S73B1JPleqfK_yEvmB3kx08JQZHrp2bJ8KQcN-kCDs">

I tried applying this and you're right, it is not working. Need to check for this case. I think it is related to the core issue about over filleting

> I understand why you don't want to add the solution in a PR in its current form but if we leave it for discussion it will be years before it is implemented, I recommend adding the PR since it is already better than what we have.
> 
> Also since Chamfers have the same exact problem do you think you can apply a similar fix to it?

For chamfers yes, it is quite simple to replicate the "Naive" class. I can do it later today.

Do anyone know if there is some documentation and/or visual explanation on how the Fillet algorithm works? I'd like to go deeper to why it can't handle over-filetting. Looks like the logic is somewhere in `ChFi3d_Builder_C1.cxx` but it is quite hard to follow...

**Comment by LinuxinaBit** [comment]
In an ideal world, would a fix for this problem make something like this possible?

https://github.com/user-attachments/assets/35f98378-3249-4402-9bed-6734abea4391

**Comment by dragazo** [comment]
@LinuxinaBit I just tested it out, and I believe the solution I proposed (doing each edge loop separately and taking the intersection) would also make what you showed in Onshape possible in FreeCAD.

In the short term if you need this feature, you can always just explicitly dupe the part, do both fillets, and manually do the intersection.

**Comment by luzpaz** [comment]
@dpasukhi would you be open to @dragazo's solution ?

**Comment by dpasukhi** [comment]
Oh, you have so much communications here. I didn't saw that issue.
I will check it.
Thank you.  If you somewhere discuss please tag me to subscribe to the messages 😅

**Comment by hyarion** [comment]
@dragazo 
> I believe the behavior used by commercial CAD software is equivalent to performing the fillet on each edge loop separately and then taking the intersection.

I don't think this is the case, but I see why you think so if you're looking at cad software using the parasolid kernel.
I made some tests to compare fusion (shapemanager/acis), onshape (parasolid), and freecad (occt) on a set of different filets. You can see the table here: https://docs.google.com/spreadsheets/d/1QyKv8SlCQUVlzpKUGoY3AiB5H8ZKX4-kI36OKvDMp8g/edit?gid=0#gid=0

As you can see in that table, there isn't any industry standard for how to handle tricky cases, in fact there's more differences than I would expect.
What's also shown is how it seems to operate behind the curtain, for example look at these where parasolid fails:

<img width="709" height="126" alt="Image" src="https://github.com/user-attachments/assets/68bd3ed7-f806-4729-a48c-9c10889d72f9" />

<img width="710" height="107" alt="Image" src="https://github.com/user-attachments/assets/bc20d024-420c-4e7f-8af9-a68a70380b8c" />

I think you're partly correct about parasolid looking at the individual edges and probably one at a time, but I don't believe they are calculating the solid for each edge filet and calculating the intersection. If they were then you wouldn't you wouldn't see the issue they have when one or more edges disappears, like in 19 in my list.

Another issue with your solution is that your solution wouldn't get rounded corners where three edges meet at a single vertex (neither internal nor external), like this:

<img width="166" height="160" alt="Image" src="https://github.com/user-attachments/assets/ffe4ee9c-de50-4f53-9b17-fc17e3a9d699" />
and

<img width="175" height="177" alt="Image" src="https://github.com/user-attachments/assets/db05a423-b398-4e45-bbab-e25be2653ae8" />

--

Slightly off topic, but notice how fusion and onshape differs in fillet philosophy for overflows.

Fusion:
* Nothing is more important than tangency for faces defined by the selected edges
* If an edge is lost we don't need to care anymore.
* If a face is collapsed, let the rolling ball roll on the resulting edge instead
* It is better with _a_ result than _no_ result

Onshape:
* Each blend is of equal importance; if one selected edge is removed, by another blend, then bail out.
* A fillet is defined by an edge where you roll a ball on the neighboring faces, if flat then extend the faces into infinite planes.
* We don't need to care about tangency between faces which shared a face that collapsed

--

I don't think there's a quick fix for occt here, and occt needs to figure out what philosophy they should follow. Personally, I would prefer onshape/parasolid's cautious approach to fail if defining edges are removed, but with fusion/shapemanager/acis' way of using face-to-edge when a face collapses into an edge:

<img width="179" height="158" alt="Image" src="https://github.com/user-attachments/assets/6ee3e0a8-b1ee-40af-a86d-771fa72882ad" />

For this issue, and other similar, the problem seems to be that there's no handling of degenerate faces. For full rounded faces, it should be possible to detect and recover from that case.

**Comment by luzpaz** [comment]
@dpasukhi what are your thoughts on https://github.com/Open-Cascade-SAS/OCCT/issues/172#issuecomment-4418443364 ?

**Comment by dpasukhi** [comment]
> [@dpasukhi](https://github.com/dpasukhi) what are your thoughts on [#172 (comment)](https://github.com/Open-Cascade-SAS/OCCT/issues/172#issuecomment-4418443364) ?

We discuss a little in Discord. That comment is helpful to know solutions and possible problems.
We will not directly implement original algorithm, but we will consolidate the logic around.

**Comment by MisterMakerNL** [comment]
@hyarion You are talking about tricky cases, your example isn't even possible to solve.
Meanwhile a simple edge consuming radius isn't possible in FreeCAD 1.1, other CAD programs you can make the radius bigger than the edge and it would still have no issues with it. This is a very basic feature that is broken in FreeCAD.

<img width="2539" height="1398" alt="Image" src="https://github.com/user-attachments/assets/efa3c3a7-cba9-42b4-8f3e-6a7e27f9abf7" />

**Comment by hyarion** [comment]
> [@hyarion](https://github.com/hyarion) You are talking about tricky cases, your example isn't even possible to solve. Meanwhile a simple edge consuming radius isn't possible in FreeCAD 1.1, other CAD programs you can make the radius bigger than the edge and it would still have no issues with it. This is a very basic feature that is broken in FreeCAD.

@MisterMakerNL What my list shows is that what you're trying to do in your screenshot isn't handled equally in other CAD kernels.

In Onshape (parasolid) you can't use that input either and on fusion (shapemanager) you get something that almost looks like a chamfer:
<img width="230" height="249" alt="Image" src="https://github.com/user-attachments/assets/d6a631ed-4ab1-4b6c-8250-2b26e773a05f" />

Dialing it down to 15mm radius so both kernels can create the blend gives you:

<img width="302" height="290" alt="Image" src="https://github.com/user-attachments/assets/56337f26-eeb8-4ce5-9cc6-e1e98390fe25" /> 

<img width="246" height="249" alt="Image" src="https://github.com/user-attachments/assets/d20b369f-8eab-45fe-8544-f5a753802d3d" />

As you can see, there isn't a standardized way to handle edge cases like this and the people behind OCCT needs to figure out what would be best for them and their clients/users. I do agree that remote blends are important but if we don't want to make this issue an epic, maybe we could just keep the original scope which I think we all agree on:

> Expected Behavior
if the fillet radius is big enough that there are no linear faces remaining, it should round the face.

**Comment by MisterMakerNL** [comment]
An edge consuming radius or fillet it's super basic and simple, it's absolutely madness that this kind of basic feature doesn't work in FreeCAD and absolutely frustrating that this isn't like priority one. There should be at least a $10k bounty on fixing this.

**Comment by FloFaber** [comment]
Fully agree with @MisterMakerNL. This is the No.1 reason for me sticking with other CAD software instead of using FreeCad.

**Comment by jhaand** [comment]
Maybe there should be an Epic on Fillets and Chamfers. Because it remains really tricky and difficult to make bulletproof.

**Comment by LinuxinaBit** [comment]
The solution here: https://github.com/Open-Cascade-SAS/OCCT/issues/172#issuecomment-3330463573 seems like a good first step. Personally I would rather have *some* of the functionality than have  *nothing* work.

That said, doing this properly is going to take some serious design work, and I don't think there's an easy way around that.

**Comment by angelobartolome** [comment]
I'd love to give it a try fixing this, I've asked previously if there is any internal documentation over the file responsible for that

> Do anyone know if there is some documentation and/or visual explanation on how the Fillet algorithm works? I'd like to go deeper to why it can't handle over-filetting. Looks like the logic is somewhere in ChFi3d_Builder_C1.cxx but it is quite hard to follow...



> An edge consuming radius or fillet it's super basic and simple, it's absolutely madness that this kind of basic feature doesn't work in FreeCAD and absolutely frustrating that this isn't like priority one. There should be at least a $10k bounty on fixing this.

**Comment by dpasukhi** [comment]
The all available documentation is what you can find in the repo:
https://dev.opencascade.org/doc/overview/html/occt_user_guides__modeling_algos.html#occt_modalg_6_1_3

That docs a little describe a logic.

From our side, I can't provide resources to solve the issues in that fast way. Unfortunately, at the moment, we do not have clients who are interested in that type of problem.
We can in some way consult developer who is interested to work on it, and provide review.

From R&D side the original task is planned, epic or subchilds will takes more time to work and will means that we starting to work on it actively - which is not true.
I only planned to spend some evenings in original task.

If someone wants to request commercial bug fixing and corrections - welcome ;) but that is big task(s) even after splitting and grouping.

**Comment by farahats9** [comment]
> From our side, I can't provide resources to solve the issues in that fast way. Unfortunately, at the moment, we do not have clients who are interested in that type of problem.

It is within your right of course to distribute your resources in a way that makes sense to your business.

However, I find it hard to believe there are no clients running into issues due to this, it is likely they are but they don't know this is the cause of it.

**Comment by MisterMakerNL** [comment]
> From our side, I can't provide resources to solve the issues in that fast way. Unfortunately, at the moment, we do not have clients who are interested in that type of problem. 

This is why FreeCAD is never going to get mainstream. It's a basic thing that is common in every other CAD program, and OCCT isn't interested in developing a CAD kernel for a CAD application. Staying with OCCT is never going to work for FreeCAD.

**Comment by dpasukhi** [comment]
> > From our side, I can't provide resources to solve the issues in that fast way. Unfortunately, at the moment, we do not have clients who are interested in that type of problem.
> 
> This is why FreeCAD is never going to get mainstream. It's a basic thing that is common in every other CAD program, and OCCT isn't interested in developing a CAD kernel for a CAD application. Staying with OCCT is never going to work for FreeCAD.

You always can develop your own kernel ;) Or you can sponsorship the development by request the commercial bug fix ;)

Or, which is more welcoming - prepare a patch and create PR with deep rework ;)

OCCT never restrict no one to use another kernels. You are welcome to use another open source kernel. Or you can use commercial kernel and pay full price ;)

**Comment by angelobartolome** [comment]
Looks like I was able to use AI to implement a fix for this (at least the initial use case  for fixing the fillet builder so a constant-radius fillet on a box edge can grow until it consumes adjacent faces and reaches the opposite edge).

I'll do some more tests.

<img width="1401" height="1110" alt="Image" src="https://github.com/user-attachments/assets/b7ea43c2-b779-4f72-9c83-43b2f467043d" />

<img width="1401" height="1110" alt="Image" src="https://github.com/user-attachments/assets/af6fb650-8741-488f-a109-fb1e78b3e1aa" />

**Comment by obelisk79** [comment]
> Looks like I was able to use AI to implement a fix for this (at least the initial use case for fixing the fillet builder so a constant-radius fillet on a box edge can grow until it consumes adjacent faces and reaches the opposite edge).


Are the changes explicit to fillet? I've long pondered if the fragility of such fillet/chamfer/shell functions was related to internal geometry handling (ie not allow null surface area, and not identifying/trimming errant geometry to maintain continuity of a surface or manifold solids). Interesting to see progress nonetheless.

**Comment by angelobartolome** [comment]
> > Looks like I was able to use AI to implement a fix for this (at least the initial use case for fixing the fillet builder so a constant-radius fillet on a box edge can grow until it consumes adjacent faces and reaches the opposite edge).
> 
> Are the changes explicit to fillet? I've long pondered if the fragility of such fillet/chamfer/shell functions was related to internal geometry handling (ie not allow null surface area, and not identifying/trimming errant geometry to maintain continuity of a surface or manifold solids). Interesting to see progress nonetheless.

I'm testing on fillet only for now, but I am very suspicious that this will apply to chamfers as well. I'll look at those later. And the helper functions can hel in shell too I guess.

WIP is here in my fork: https://github.com/Open-Cascade-SAS/OCCT/compare/master...angelobartolome:OCCT:fix/fillet

<img width="718" height="416" alt="Image" src="https://github.com/user-attachments/assets/8c025470-df5e-4fc0-b864-1bc72def9491" />

<img width="462" height="403" alt="Image" src="https://github.com/user-attachments/assets/63eeb91c-5536-4387-a340-45fd4e591748" />

I think this is a huge step already

**Comment by jenkitay** [comment]
> Looks like I was able to use AI to implement a fix for this (at least the initial use case for fixing the fillet builder so a constant-radius fillet on a box edge can grow until it consumes adjacent faces and reaches the opposite edge).
> 
Awesome! Thank you!

Reading through this issue, it seems that much of the holdup has been trying to decide what the preferred proper behavior should be. It seems that you have certainly achieved at least one of the possible proper behaviors. If multiple ways of handling edge cases like this exist, then perhaps instead of the developer deciding how it should be handled, give the user options to decide how they want it to be handled. Nevertheless, if the developer can only handle it one way, that's better than no way.

**Comment by jhaand** [comment]
@jenkitay   
> Reading through this issue, it seems that much of the holdup has been trying to decide what the preferred proper behavior should be. It seems that you have certainly achieved at least one of the possible proper behaviors.

I find it strange that the 3D CAD industry can come up with a standard like STEP but can't describe the proper behaviour and requirements for fillets and chamfers. At least the code now gets a cleanup and a good proposal. 

Great work picking this up @angelobartolome .

**Comment by MisterMakerNL** [comment]
If we can fix that, 

> > > From our side, I can't provide resources to solve the issues in that fast way. Unfortunately, at the moment, we do not have clients who are interested in that type of problem.
> > 
> > 
> > This is why FreeCAD is never going to get mainstream. It's a basic thing that is common in every other CAD program, and OCCT isn't interested in developing a CAD kernel for a CAD application. Staying with OCCT is never going to work for FreeCAD.
> 
> You always can develop your own kernel ;) Or you can sponsorship the development by request the commercial bug fix ;)
> 
> Or, which is more welcoming - prepare a patch and create PR with deep rework ;)
> 
> OCCT never restrict no one to use another kernels. You are welcome to use another open source kernel. Or you can use commercial kernel and pay full price ;)

A billion dollar-company asking for sponsorship from a Opensource organization.... Come on man that's bs and you know it.

Your org. could even sponsor Freecad get a tax refund, FreeCAD pays you money back = free money. Although I think FPA didn't registered as a charity 😭.

Gladly we got some awesome people in the community @angelobartolome ! Hope it'll work!

**Comment by luzpaz** [comment]
@MisterMakerNL, @dpasukhi has been nothing but kind and very attentive. I don't think he should be the target of angry rants or hot takes. Feel free to post that on the opencascade forum or freecad forum. Opining here is misplaced and is only going to derail further civil cooperation.

**Comment by MisterMakerNL** [comment]
I won't go further into this but he/she/them is representing OCCT/Capgemini and so I speak true them to that organization.
I assume he understands it's not personal.
And as somone who worked for big corp, he/she/them probably understands it very well.

**Comment by luzpaz** [comment]
Thank you for working on this 🙏

**Comment by dpasukhi** [comment]
Thank you for bug reports and special thanks to @angelobartolome to the initiating development process and showing that fix is possible ;)

Fix itself is merged to IR branch. Some post-merge PR can be arrive before merging to master.

The changes is covered reported issues. The shape modification history is preserved.
According OCCT release circles, the changes will be only available in 8.1.0 official release and as a result, FreeCAD stable release can use these changes only after that. Approximation of 8.1.0 release can be checked here: https://github.com/Open-Cascade-SAS/OCCT/milestones

In case of other issues for chamfers/fillets, please consult with Pieter (if FreeCAD related) and report tickets to FreeCAD project or OCCT in case if you have C++/DRAW code to reproduce.
